rust_src/src/eval.rs

```diff
@@ -1,91 +1,100 @@
-            )
+/// Eval args until one of them yields non-nil, then return that value.
+/// The remaining args are not evalled at all.
+/// If all args return nil, return nil.
+/// usage: (or CONDITIONS...)
+#[lisp_fn(min = "0", unevalled = "true")]
+pub fn or(args: LispObject) -> LispObject {
+    let mut val = Qnil;
+
+    for elt in args.iter_cars_safe() {
+        val = unsafe { eval_sub(elt.to_raw()) };
+        if val != Qnil {
+            break;
         }
-    }}
+    }
+
+    LispObject::from_raw(val)
 }
 
-macro_rules! message_with_string {
-    ($str:expr, $obj:expr, $should_log:expr) => {
-        unsafe {
-            ::remacs_sys::message_with_string($str.as_ptr() as *const ::libc::c_char,
-                                              $obj.to_raw(),
-                                              $should_log);
+/// Eval args until one of them yields nil, then return nil.
+/// The remaining args are not evalled at all.
+/// If no arg yields nil, return the last arg's value.
+/// usage: (and CONDITIONS...)
+#[lisp_fn(min = "0", unevalled = "true")]
+pub fn and(args: LispObject) -> LispObject {
+    let mut val = Qnil;
+
+    for elt in args.iter_cars_safe() {
+        val = unsafe { eval_sub(elt.to_raw()) };
+        if val == Qnil {
+            break;
         }
     }
-}
 
-/// Macro to format an error message.
-/// Replaces error() in the C layer.
-macro_rules! error {
-    ($str:expr) => {{
-        let strobj = unsafe {
-            ::remacs_sys::make_string($str.as_ptr() as *const ::libc::c_char,
-                                      $str.len() as ::libc::ptrdiff_t)
-        };
-        xsignal!(::remacs_sys::Qerror, $crate::lisp::LispObject::from_raw(strobj));
-    }};
-    ($fmtstr:expr, $($arg:expr),*) => {{
-        let formatted = format!($fmtstr, $($arg),*);
-        let strobj = unsafe {
-            ::remacs_sys::make_string(formatted.as_ptr() as *const ::libc::c_char,
-                                      formatted.len() as ::libc::ptrdiff_t)
-        };
-        xsignal!(::remacs_sys::Qerror, $crate::lisp::LispObject::from_raw(strobj));
-    }};
+    LispObject::from_raw(val)
 }
 
-/// Macro to format a "wrong argument type" error message.
-macro_rules! wrong_type {
-    ($pred:expr, $arg:expr) => {
-        xsignal!(::remacs_sys::Qwrong_type_argument, LispObject::from_raw($pred), $arg);
-    };
+/// If COND yields non-nil, do THEN, else do ELSE...
+/// Returns the value of THEN or the value of the last of the ELSE's.
+/// THEN must be one expression, but ELSE... can be zero or more expressions.
+/// If COND yields nil, and there are no ELSE's, the value is nil.
+/// usage: (if COND THEN ELSE...)
+#[lisp_fn(name = "if", c_name = "if", min = "2", unevalled = "true")]
+pub fn lisp_if(args: LispObject) -> LispObject {
+    let cell = args.as_cons_or_error();
+    let cond = unsafe { eval_sub(cell.car().to_raw()) };
+
+    if cond == Qnil {
```

Comment:
This is putting the ELSE before the THEN. I think it's clearer with THEN first, which is what the C does.

Reply:
Will do